p2p/net/network.go

```diff
@@ -323,47 +298,58 @@ func (n *Net) publishNewRemoteConnectionEvent(conn Connection, node node.Node) {
 
 // HandlePreSessionIncomingMessage establishes session with the remote peer and update the Connection with the new session
 func (n *Net) HandlePreSessionIncomingMessage(c Connection, message []byte) error {
-	//TODO replace the next few lines with a way to validate that the message is a handshake request based on the message metadata
-	errMsg := "failed to handle handshake request"
-	data := &pb.HandshakeData{}
-
-	err := proto.Unmarshal(message, data)
+	message, remotePubkey, err := cryptoBox.ExtractPubkey(message)
 	if err != nil {
-		return fmt.Errorf("%s. err: %v", errMsg, err)
+		return err
 	}
+	c.SetRemotePublicKey(remotePubkey)
+	session := createSession(n.localNode.PrivateKey(), remotePubkey)
+	c.SetSession(session)
 
-	// new remote connection doesn't hold the remote public key until it gets the handshake request
-	if c.RemotePublicKey() == nil {
-		rPub, err := crypto.NewPublicKey(data.GetNodePubKey())
-		n.Logger().Debug("DEBUG: handling HS req from %v", rPub)
-		if err != nil {
-			return fmt.Errorf("%s. err: %v", errMsg, err)
-		}
-		c.SetRemotePublicKey(rPub)
-	}
-	respData, session, err := ProcessHandshakeRequest(n.NetworkID(), n.localNode.PublicKey(), n.localNode.PrivateKey(), c.RemotePublicKey(), data)
-	if err != nil {
-		return fmt.Errorf("%s. err: %v", errMsg, err)
-	}
-	payload, err := proto.Marshal(respData)
+	// open message
+	protoMessage, err := session.OpenMessage(message)
 	if err != nil {
-		return fmt.Errorf("%s. hereto err: %v", errMsg, err)
+		return err
 	}
 
-	// update on new connection
-	addr, _, err := net.SplitHostPort(c.RemoteAddr().String())
+	handshakeData := &pb.HandshakeData{}
+	err = proto.Unmarshal(protoMessage, handshakeData)
 	if err != nil {
-		return fmt.Errorf("un-valid address format, (%v) err: %v", c.RemoteAddr().String(), err)
+		return err
 	}
 
-	anode := node.New(c.RemotePublicKey(), net.JoinHostPort(addr, strconv.Itoa(int(data.Port))))
-
-	err = c.Send(payload)
+	remoteListeningPort := uint16(handshakeData.Port)
+	remoteListeningAddress, err := replacePort(c.RemoteAddr().String(), remoteListeningPort)
 	if err != nil {
 		return err
 	}
+	anode := node.New(c.RemotePublicKey(), remoteListeningAddress)
 
-	c.SetSession(session)
 	n.publishNewRemoteConnectionEvent(c, anode)
+	// TODO: @noam process message?
```

Comment:
open an issue about this, we used to check networkid and clientversion with `ProcessHandshakeRequest`

Reply:
I've implemented it when receiving connections (used the code from before this change) and added an issue for verifying peers we are dialing to: https://github.com/spacemeshos/go-spacemesh/issues/472.